Moving the server-docker repository to the server repository and enabling image building from the current version of the application code (#48)
getferdi/server: docker/README.md

The docker can be run as is, with the default sqlite database, or you can modify your environment variables to use an external database (e.g. MySQL, MariaDB, Postgres, etc). After setting up the docker container you will likely need to create a reverse proxy to access Ferdi-server outside of your home network, using a webserver such as NGINX.

### docker

Pull the docker image:

    docker pull getferdi/ferdi-server

To create the docker container with the proper parameters:

	docker create \
	  --name=ferdi-server \
	  -e NODE_ENV=development \
	  -e EXTERNAL_DOMAIN=<ferdi-serverdomain> \  
	  -e DB_CONNECTION=<database> \
	  -e DB_HOST=<yourdbhost> \
	  -e DB_PORT=<yourdbport> \
	  -e DB_USER=<yourdbuser> \
	  -e DB_PASSWORD=<yourdbpass> \
	  -e DB_DATABASE=<yourdbdatabase> \
	  -e DB_SSL=false \
	  -e MAIL_CONNECTION=smtp \
	  -e SMPT_HOST=<smtpmailserver> \
	  -e SMTP_PORT=<smtpport> \
	  -e MAIL_SSL=true/false \
	  -e MAIL_USERNAME=<yourmailusername> \
	  -e MAIL_PASSWORD=<yourmailpassword> \
	  -e MAIL_SENDER=<sendemailaddress> \
	  -e IS_CREATION_ENABLED=true \
	  -e IS_DASHBOARD_ENABLED=true \
	  -e IS_REGISTRATION_ENABLED=true \
	  -e CONNECT_WITH_FRANZ=true \

> 	  -e DATA_DIR=data \

Reviewer: ```suggestion
	  -e DATA_DIR=/data \
```

> 	  -p <port>:80 \

Reviewer: If I remember correctly the port is 3333 by default, right?

```suggestion
	  -p <port>:3333 \
```

> 	  -v <path to data>:/config \
> 	  -v <path to database>:/app/database \
> 	  -v <path to recipes>:/app/recipes \

Reviewer: ```suggestion
	  -v <path to data>:/data \
	  -v <path to recipes>:/app/recipes \
```

	  --restart unless-stopped \
	  getferdi/ferdi-server

### docker-compose

  You can use the provided sample [./docker/docker-compose.yml](https://github.com/getferdi/server/tree/master/docker/docker-compose.yml) if you are happy with the default environmental variables. This will pull the latest image from Docker Hub or use a local copy of the image which you can build using the instructions provided in the [Building locally section](#building-locally).
  
  To start the application, use
		
		docker-compose up
The server will be launched at [http://localhost:3333/](http://localhost:3333/)	address.

> **Existing users please note:** The latest updates to Ferdi-server and the Ferdi-server Docker image introduce changes to the default sqlite database name and location, please see the comments in the sample [./docker/docker-compose.yml](https://github.com/getferdi/server/tree/master/docker/docker-compose.yml) in order to continue using your existing database.

Reviewer 2: Please review the warning for existing users here and on lines 78-79. I have also corrected the spelling and grammar throughout the README, so please check that everything is correct.

Reviewer: The warning sold be in the main readme, shouldn't it? Because we changed it for everyone, not just docker users.

Reviewer 2: Maybe both....

Reviewer: Well, sure

Reviewer 2: Added a warning to the main readme and pointed existing users to the docker documentation for  more info.

## Configuration

Container images are configured using parameters passed at runtime (such as those above). 

**Warning, the use of `config.txt` is now deprecated. Please make sure to pass the correct environmental variables to your container at runtime.** 
<strike>If any environmental parameter is not passed to the container, its value will be taken from the `/config/config.txt` file.</strike> 

| Parameter | Function |
| :----: | --- |

> | `-p <port>:3333` | Will map the container's port 3333 to a port on the host, default is 3333 |

Reviewer 2: Please ensure that this is correct - it was still listed as mapping host port 3333 to container port 80, which I believe is no longer correct.

Reviewer: Maybe we should just state the default port in the docker container here and refer to the docker docs for mapping?

Reviewer: (same for the variables, describe the in the main README and like that here along with a link on how to do it in docker)

Reviewer 2: Please check my additions, but I think I've resolved this.

| `-e SMPT_HOST=<smtpmailserver>` | for specifying the mail host to be used, default is 127.0.0.1 |
| `-e SMTP_PORT=<smtpport>` | for specifying the mail port to be used, default is 2525 |
| `-e MAIL_SSL=true/false` | for specifying SMTP mail security, default is false |
| `-e MAIL_USERNAME=<yourmailusername>` | for specifying your mail username to be used, default is username |
| `-e MAIL_PASSWORD=<yourmailpassword>` | for specifying your mail password to be used, default is password |
| `-e MAIL_SENDER=<sendemailaddress` | for specifying the mail sender address to be used, default is noreply@getferdi.com |
| `-e IS_CREATION_ENABLED=true` | for specifying whether to enable the [creation of custom recipes](#creating-and-using-custom-recipes), default is true |
| `-e IS_DASHBOARD_ENABLED=true` | for specifying whether to enable the Ferdi-server dashboard, default is true |
| `-e IS_REGISTRATION_ENABLED=true` | for specifying whether to allow user registration, default is true |
| `-e CONNECT_WITH_FRANZ=true` | for specifying whether to enable connections to the Franz server, default is true |
| `-e DATA_DIR=data` | for specifying the SQLite database folder, default is data |

> | `-v <path to data>:/config` | this will store persistent configuration data on the docker host. See the [Docker docs](https://docs.docker.com/storage/volumes/) for more information on the use of container volumes |
> | `-v <path to database>:/app/data` | this will store Ferdi-server's database on the docker host for persistence |
> | `-v <path to recipes>:/app/recipes` | this will store Ferdi-server's recipes on the docker host for persistence |

Reviewer: ```suggestion
| `-v <path to data>:/data` | this will store Ferdi-server's data (among others the database) on the docker host for persistence. See the [Docker docs](https://docs.docker.com/storage/volumes/) for more information on the use of container volumes |
| `-v <path to recipes>:/app/recipes` | this will store Ferdi-server's recipes on the docker host for persistence |
```

> <p align="center">
>     <img src="./logo.png" alt="" width="300"/>  
> </p>
> 

Reviewer: ```suggestion
```

Reviewer: I think we can remove this, as it's now just some documentation inside of the repo and we probably don't need the logo file either

# Ferdi-server-docker
[Ferdi](https://github.com/getferdi/ferdi) is a hard-fork of [Franz](https://github.com/meetfranz/franz), adding awesome features and removing unwanted ones. Ferdi-server is an unofficial replacement of the Franz server for use with the Ferdi Client. 

This is a dockerized version of [Ferdi-server](https://github.com/getferdi/server) running on Alpine Linux and Node.js (v10.16.3).

## Why use a custom Ferdi-server?
A custom ferdi-server allows you to experience the full potential of the Ferdi client. It allows you to use all Premium features (e.g. Workspaces and custom URL recipes) and [adding your own recipes](#creating-and-using-custom-recipes).

## Features
- [x] User registration and login
- [x] Service creation, download, listing and removing
- [x] Workspace support
- [x] Functioning service store
- [x] User dashboard
- [x] Password recovery
- [x] Export/import data to other ferdi-servers
- [ ] Recipe update

## Installation & Setup

Here are some example snippets to help you get started creating a container.

> The docker can be run as is, with the default sqlite database, or you can modifying your ENV variables to use an external database (e.g. MYSql, MariaDB, Postgres, etc). After setting up the docker container you will need to create a NGINX reverse proxy to access Ferdi-server outside of your home network. 

Reviewer: ```suggestion
The docker can be run as is, with the default sqlite database, or you can modify your environment variables to use an external database (e.g. MySQL, MariaDB, Postgres, etc). After setting up the docker container you will likely need to create a reverse proxy to access Ferdi-server outside of your home network, using for example NGINX.
```

Reviewer: This makes some strong assumptions, I edited it a bit to reflect the nature of those recommendations/examples

Author: true, I will fix that.

### docker

Pull the docker image:

    docker pull getferdi/ferdi-server

To create the docker container with the proper parameters:

> 	docker create \

Reviewer: Maybe add `-e DATA_DIR=data here? (See above)

	  -e DATA_DIR=data \
	  -p <port>:80 \
	  -v <path to data>:/config \
	  -v <path to database>:/app/database \
	  -v <path to recipes>:/app/recipes \
	  --restart unless-stopped \
	  getferdi/ferdi-server

### docker-compose

  You can use sample [./docker/docker-compose.yml](https://github.com/getferdi/server/tree/master/docker/docker-compose.yml).
  This will pull latest image from Docker Hub or use local image which you can build using instructions in [Building locally section](#Building-locally).
  
  To start the application, use
		
		docker-compose up
The server will be launched at [http://localhost:3333/](http://localhost:3333/)	address.

## Configuration

Container images are configured using parameters passed at runtime (such as those above). These parameters are separated by a colon and indicate `<external>:<internal>` respectively. For example, `-p 3333:80` would expose port `80` from inside the container to be accessible from the host's IP on port `3333` outside the container. 
After the first run, Ferdi-server's default configuration is saved inside the `config.txt` file inside your persistent data directory (`/config` in the container).

> If any environmental parameter is not passed to the container, its value will be taken from the `config.txt` file. You can also edit the `config.txt` file, but it is advisable to send the environment parameters to the container.

Reviewer: Maybe remove it from here or add a deprecation notice?

| `-e SMPT_HOST=<smtpmailserver>` | for specifying the mail host to be used, default is 127.0.0.1 |
| `-e SMTP_PORT=<smtpport>` | for specifying the mail port to be used, default is 2525 |
| `-e MAIL_SSL=true/false` | for specifying SMTP mail secuirty, default is false |
| `-e MAIL_USERNAME=<yourmailusername>` | for specifying your mail username to be used, default is username |
| `-e MAIL_PASSWORD=<yourmailpassword>` | for specifying your mail password to be used, default is password |
| `-e MAIL_SENDER=<sendemailaddress` | for specifying the mail sender address to be used, default is noreply@getferdi.com |
| `-e IS_CREATION_ENABLED=true` | for specifying whether to enable the [creation of custom recipes](#creating-and-using-custom-recipes), default is true |
| `-e IS_DASHBOARD_ENABLED=true` | for specifying whether to enable the Ferdi-server dashboard, default is true |
| `-e IS_REGISTRATION_ENABLED=true` | for specifying whether to allow user registration, default is true |
| `-e CONNECT_WITH_FRANZ=true` | for specifying whether to enable connections to the Franz server, default is true |

> | `-e DATA_DIR=database` | for specifying sql-lite database folder, default is database. Currently you need to use folder inside `/app` directory |

Reviewer: Same no need for compatibility in the recommendation

| `-e DATA_DIR=data` | for specifying sql-lite database folder, default is database. |
| `-v <path to data>:/config` | this will store persistent ENV  data on the docker host |
| `-v <path to database>:/app/database` | this will strore Ferdi-server's database on the docker host for persistence |
| `-v <path to recipes>:/app/recipes` | this will strore Ferdi-server's recipes on the docker host for persistence |

By enabling the `CONNECT_WITH_FRANZ` option, Ferdi-server can:
    - Show the full Franz recipe library instead of only custom recipes
    - Import Franz accounts

## Supported databases and drivers

To use a different database than the default, SQLite, enter the driver code below in your ENV configuration. 

| Database | Driver |
| :----: | --- |

> | MariaDB | mysql |
> | MySQL | mysql |

Reviewer: ```suggestion
| MariaDB/MySQL | mysql |
```

Reviewer: Just some clutter that's not particularly bad, but you know 🤷

| PostgreSQL | pg |
| SQLite3 | sqlite |

## Supported mail connections (advanced)

To use a different email sender than the default, SMTP, enter the correct information in your ENV configuration and adapt your docker run, create, or compose commands accordingly.

| Mail Connection | ENV variables |
| :----: | --- |
| SMTP | SMTP_PORT, SMTP_HOST, MAIL_USERNAME, MAIL_PASSWORD, MAIL_SSL |
| SparkPost | SPARKPOST_API_KEY |
| Mailgun | MAILGUN_DOMAIN, MAILGUN_API_REGION, MAILGUN_API_KEY |

> | Ethereal | A disposable account is created automatically if you choose this option. |

Reviewer: The new adonis version doesn't support this anymore, so maybe we should add a deprecation notice here?

Author: If so then imho we should add this information.

Author: Fixed in [https://github.com/getferdi/server/pull/48/commits/88c94fbb301f7778f288d08e6daa81bef96a7e7f](https://github.com/getferdi/server/pull/48/commits/88c94fbb301f7778f288d08e6daa81bef96a7e7f)

## Transferring user data
Please refer to <https://github.com/getferdi/ferdi/wiki/Transferring-data-between-servers>

## Creating and using custom recipes
Ferdi-server allows to extends the Franz recipe catalogue with custom Ferdi recipes.

For documentation on how to create a recipe, please visit [the official guide by Franz](https://github.com/meetfranz/plugins/blob/master/docs/integration.md).

To add your recipe to Ferdi-server, open `http://[YOUR FERDI-SERVER]/new` in your browser. You can now define the following settings:
- `Author`: Author who created the recipe
- `Name`: Name for your new service. Can contain spaces and unicode characters
- `Service ID`: Unique ID for this recipe. Does not contain spaces or special characters (e.g. `google-drive`)
- `Link to PNG/SVG image`: Direct link to a 1024x1024 PNG image and SVG that is used as a logo inside the store. Please use jsDelivr when using a file uploaded to GitHub as raw.githubusercontent files won't load

> - `Recipe files`: Recipe files that you created using the [Franz recipe creation guide](https://github.com/meetfranz/plugins/blob/master/docs/integration.md). Please do *not* package your files beforehand - upload the raw files (you can drag and drop multiple files). ferdi-server will automatically package and store the recipe in the right format. Please also do not drag and drop or select the whole folder, select the individual files.

Reviewer: I also think this might be not the right, place but maybe something to fix after this PR

Author: I leave that as it is. I don't have enough info about it to make the change.

* Recreate a new container with the same docker create parameters as instructed above (if mapped correctly to a host folder, your `/config` folder and your ENV settings will be preserved)
* Start the new container: `docker start ferdi-server`
* You can also remove the old dangling images: `docker image prune`

### Via Docker Compose
* Update all images: `docker-compose pull`
  * or update a single image: `docker-compose pull ferdi-server`
* Let compose update all containers as necessary: `docker-compose up -d`
  * or update a single container: `docker-compose up -d ferdi-server`
* You can also remove the old dangling images: `docker image prune`

## Building locally

If you want to build this image locally, please run this command from root of [ferdi server repository](https://github.com/getferdi/server/tree/master/):
```
docker build \
  --no-cache \
  --pull \
  -t getferdi/ferdi-server:latest .
```

## Versions

* **05.02.21:** - Repository moved to ferdi-server repository
* **19.01.21:** - Updated Mail SSL and DB SLL settings
* **20.09.20:** - Updated SMTP Mailer settings for password reset.
* **21.06.20:** - Rebase to Alpine 3.11 and added Mailer settings.

> * **25.09.19:** - Initial Release.

Reviewer: I think this can also go away, if we want to release real tagged versions?

Author: Fixed in [https://github.com/getferdi/server/commit/0a3ce5a9e37dd917cfabdf97fdc18a41d88c65eb](https://github.com/getferdi/server/commit/0a3ce5a9e37dd917cfabdf97fdc18a41d88c65eb)

This is a dockerized version of [Ferdi-server](https://github.com/getferdi/server) running on Alpine Linux and Node.js (v10.16.3).

## Why use a custom Ferdi-server?
A custom Ferdi-server allows you to experience the full potential of the Ferdi Client. It allows you to use all Premium features (e.g. Workspaces and custom URL recipes) and [add your own recipes](#creating-and-using-custom-recipes).

## Features
- [x] User registration and login
- [x] Service creation, download, listing and removing
- [x] Workspace support
- [x] Functioning service store
- [x] User dashboard
- [x] Password recovery
- [x] Export/import data to other Ferdi-servers
- [ ] Recipe update

## Installation & Setup

Here are some example snippets to help you get started creating a container.

The docker can be run as is, with the default sqlite database, or you can modify your environment variables to use an external database (e.g. MySQL, MariaDB, Postgres, etc). After setting up the docker container you will need to create a reverse proxy to access Ferdi-server outside of your home network, using a webserver such as NGINX.

> **Existing users, please note:** The latest updates to Ferdi-server and the Ferdi-server Docker image introduce changes to the default SQLite database name and location, as well as the internal container port. The new container port is `3333`. If you would like to keep your existing SQLite database, you will need to add the `DATA_DIR` variable and change it to `/app/database`, to match your existing data volume. You will also need to change the `DB_DATABASE` variable to `development` to match your existing database. Please see the parameters in the [Migration section](#migrating-from-an-existing-ferdi-server) below.

Reviewer: Well, that's the thing... There's no versions until now... It's all `latest`.

> **Existing users, please note:** The latest updates to Ferdi-server and the Ferdi-server Docker image introduce changes to the default SQLite database name and location, as well as the internal container port. The new container port is `3333`. If you would like to keep your existing SQLite database, you will need to add the `DATA_DIR` variable and change it to `/app/database`, to match your existing data volume. You will also need to change the `DB_DATABASE` variable to `development` to match your existing database. Please see the parameters in the [Migration section](#migrating-from-an-existing-ferdi-server) below.

Reviewer 2: That's part of the reason I want to get this merged soon, even as a beta/pre-release, in order to get proper versioning in place.

| `-e IS_CREATION_ENABLED=true` | for specifying whether to enable the [creation of custom recipes](#creating-and-using-custom-recipes), default is true |
| `-e IS_DASHBOARD_ENABLED=true` | for specifying whether to enable the Ferdi-server dashboard, default is true |
| `-e IS_REGISTRATION_ENABLED=true` | for specifying whether to allow user registration, default is true |
| `-e CONNECT_WITH_FRANZ=true` | for specifying whether to enable connections to the Franz server, default is true |
| `-e DATA_DIR=data` | for specifying the SQLite database folder, default is data |
| `-v <path to data on host>:/data` | this will store Ferdi-server's data (its database, among other things) on the docker host for persistence. See the [Docker docs](https://docs.docker.com/storage/volumes/) for more information on the use of container volumes |
| `-v <path to recipes on host>:/app/recipes` | this will store Ferdi-server's recipes on the docker host for persistence |

By enabling the `CONNECT_WITH_FRANZ` option, Ferdi-server can:
    - Show the full Franz recipe library instead of only custom recipes
    - Import Franz accounts

## Supported databases and drivers

> To use a different database than the default, SQLite3, enter the driver code below in your ENV configuration. 

Reviewer: Nope, no need for anything OS, just change the env variable at least that seems to be working fine for PostgreSQL and I am quite sure it's the same for MySQL.

| Parameter | Function |
| :----: | --- |
| `-e DB_CONNECTION=mysql` | for specifying the database being used |
| `-e DB_HOST=192.168.10.1` | for specifying the database host machine IP |
| `-e DB_PORT=3306` | for specifying the database port |
| `-e DB_USER=ferdi` | for specifying the database user |
| `-e DB_PASSWORD=ferdipw` | for specifying the database password|
| `-e DB_DATABASE=adonis` | for specifying the database to be used|
| `-v <path to database>:/app/database` | this will strore Ferdi-server's database on the docker host for persistence |
| `-v <path to recipes>:/app/recipes` | this will strore Ferdi-server's recipes on the docker host for persistence |

**In eithr case, pleasee be sure to pass the correct variables to the new Ferdi-server container in order maintain access to your existing database.**
 
## NGINX config block

To access Ferdi-server from outside of your home network on a subdomain use this server block:

```
# Ferdi-server
server {
        listen 443 ssl http2;
        server_name ferdi.my.website;

>        	# all ssl related config moved to ssl.conf

Reviewer 3: indentation

Reviewer 2: I believe this should be fixed now.

## Configuration

Container images are configured using parameters passed at runtime (such as those above). 

> **Warning ,using `config.txt` will be deprecated in the future releases.** 

Reviewer: ```suggestion
**Warning, using `config.txt` will be deprecated in the future releases.** 
```
